src/lint/track_config.nim

```diff
@@ -388,6 +389,43 @@ proc checkExercisesPCP(exercises: seq[ConceptExercise] | seq[PracticeExercise];
               "Exercise is allowed to have that"
       b.setFalseAndPrint(msg, path)
 
+proc checkExerciseSlugsAndForegone(trackConfig: TrackConfig; b: var bool;
+                                   path: Path) =
+  ## Sets `b` to `false` if the below conditions are not satisfied:
+  ## - Each slug of a Concept Exercise or a Practice Exercise in `trackConfig`
+  ##   only exists once on the track.
+  ## - There is exactly one Practice Exercise with the slug `hello-world`.
+  ## - The `foregone` array does not contain a slug of an implemented exercise.
+  var conceptExerciseSlugs = initHashSet[string](200)
```

Comment:
Is the `200` just an initial size, or is it a maximum?

Reply:
It's an initial size - how many items the `HashSet` should support. But, like Nim sequences, a `HashSet` "expands" automatically if you keep adding items.

---

In case you're interested, the implementation details for `HashSet` are here:
- [`lib/pure/collections/sets.nim`](https://github.com/nim-lang/Nim/blob/ee2eb5cae258/lib/pure/collections/sets.nim#L58-L67)
- [`lib/pure/collections/setimpl.nim`](https://github.com/nim-lang/Nim/blob/ee2eb5cae258/lib/pure/collections/setimpl.nim#L18-L24)
- [`lib/pure/collections/hashcommon.nim`](https://github.com/nim-lang/Nim/blob/ee2eb5cae258/lib/pure/collections/hashcommon.nim#L10-L34)

We allocate an initial number of "slots" using the formula `theIntegerThatYouPass * 3 div 2 + 4`, and in general the HashSet expands when it's more than two-thirds full.

And as a quick demonstration, let's use `initHashSet(3)` (8 slots, by the above calculation) and then insert 5 items,  print the representation, then add another item, and print the representation agin:

```Nim
import std/sets

var s = initHashSet[string](3)
s.incl "aaa"
s.incl "bbb"
s.incl "ccc"
s.incl "ddd"
s.incl "eee"
echo s.repr
echo "--------------"
# We rehash when `(t.dataLen * 2 < t.counter * 3) or (t.dataLen - t.counter < 4)`
# An `initialSize` of 3 produces a `dataLen` of 8 , so the above condition becomes true when `t.counter` is 5
# Basically, we rehash when the HashSet is more than two thirds full.
s.incl "fff"
echo s.repr
```

The output looks something like the below. We can see that:
- the `HashSet` grows as we insert the 6th item
- the resulting HashSet occupies double the memory (because `growthFactor == 2`)
- (if you stare hard enough) the pointer to the `HashSet` changes - we allocated a new block of memory

```
[data = 0x7fac8a41c050@[
  [Field0 = 735026264, Field1 = 0x7fac8a41d0c0"ccc"],
  [Field0 = 2446311049, Field1 = 0x7fac8a41d090"bbb"],
  [Field0 = 4095589471, Field1 = 0x7fac8a41d0f0"ddd"],
  [Field0 = 1505603105, Field1 = 0x7fac8a41d120"eee"],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 3033554871, Field1 = 0x7fac8a41d060"aaa"]],
counter = 5]

--------------

[data = 0x7fac8a41e050@[
  [Field0 = 966543936, Field1 = 0x7fac8a41dab0"fff"],
  [Field0 = 1505603105, Field1 = 0x7fac8a41da50"eee"],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 3033554871, Field1 = 0x7fac8a41da80"aaa"],
  [Field0 = 735026264, Field1 = 0x7fac8a41d9c0"ccc"],
  [Field0 = 2446311049, Field1 = 0x7fac8a41d9f0"bbb"],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 0, Field1 = ""],
  [Field0 = 4095589471, Field1 = 0x7fac8a41da20"ddd"]],
counter = 6]
```

Needless allocations are an important source of performance problems, so it's a good optimization to avoid growing strings/seqs/HashSets etc when they're large and you're doing a lot of work. But of course, it doesn't matter at all in our case.

Reply:
Cool, thanks.